File: CSharp/ByTier/B/2029_B_Replacement.cs
// Problem: CF 2029 B - Replacement
// https://codeforces.com/contest/2029/problem/B

using System;
public class Replacement {
  // Purpose: Determine if a binary string s can be reduced to a single character
  //          using a sequence of operations defined by string r.
  //          Each operation selects a position k where s[k] != s[k+1] and replaces
  //          s[k]s[k+1] with r[i]. 
  //          The key insight is that we must ensure we always have a valid k to operate on.
  //          Time Complexity: O(n) per test case
  //          Space Complexity: O(1) - only using a few variables
  
  public static void Main() {
    short t = short.Parse(Console.ReadLine());
    for (short i = 0; i < t; i++) {
      int n = int.Parse(Console.ReadLine());
      string s = Console.ReadLine();
      string r = Console.ReadLine();
      int count0 = 0;
      int count1 = 0;
      // Count total number of 0s and 1s in s
      for (int j = 0; j < n; j++) {
        if (s[j] == '0')
          count0++;
        else
          count1++;
      }
      int idxLast = n - 1;
      bool ans = true;
      // Process each operation in r
      for (int j = 0; j < idxLast && ans; j++) {
        // If no more 0s or 1s, we can't make a move
        if (count0 == 0 || count1 == 0)
          ans = false;
        else if (r[j] == '0')
          count1--; // We are replacing a pair with a '0', so decrement 1s
        else
          count0--; // We are replacing a pair with a '1', so decrement 0s
      }
      Console.WriteLine(ans ? "YES" : "NO");
    }
  }
}

// https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/B/2029_B_Replacement.cpp
// Problem: CF 2029 B - Replacement
// https://codeforces.com/contest/2029/problem/B

/*
B. Replacement
Algorithm: Greedy simulation with character count tracking
Time Complexity: O(n) per test case, where n is the length of string s
Space Complexity: O(1) additional space (not counting input storage)

The problem involves simulating a game where we reduce a binary string s of length n to a single character
by performing n-1 operations. In each operation, we must choose an adjacent pair of different characters,
and replace the pair with a character from r. The goal is to determine if it's possible to perform all operations.

Approach:
- Count the number of 0s and 1s in s.
- For each step in simulation, check if we can continue:
  - We need at least one 0 and one 1 to make a valid operation.
  - Based on r[i], we decrease the count of either 0 or 1.
- If at any point we run out of valid pairs (e.g., all 0s or all 1s), return false.
*/

#include <iostream>
#include <string>

int main() {
  std::ios_base::sync_with_stdio(false);
  long t;
  std::cin >> t;
  while (t--) {
    long n;
    std::cin >> n;
    std::string s;
    std::cin >> s;
    std::string r;
    std::cin >> r;
    long a(0), z(0); // Count of '1's and '0's in s
    for (long p = 0; p < s.size(); p++) {
      z += (s[p] == '0'); // Count number of zeros
      a += (s[p] == '1'); // Count number of ones
    }
    bool res(true); // Whether we can complete all operations
    for (long p = 0; p < n - 1 && res; p++) {
      if (!z || !a) { // If there's no valid adjacent pair left
        res = false;
      }
      if (r[p] == '0') { // If replacement character is '0', we decrease count of '1's
        --a;
      } else { // Else, decrease count of '0's
        --z;
      }
    }
    std::cout << (res ? "YES" : "NO") << std::endl;
  }
}


// https://github.com/VaHiX/codeForces/